container/container.go

```diff
@@ -803,7 +803,7 @@ type rio struct {
 func (i *rio) Close() error {
 	i.IO.Close()
 
-	return i.sc.CloseStreams()
+	return i.sc.CloseStreams(context.Background())
```

Comment:
We should probably have a default timeout here – otherwise, cases like https://github.com/moby/moby/issues/46078#issuecomment-1667840537

where the blocking call to `CloseStreams` comes from
```
goroutine 1174391 [sync.Mutex.Lock, 508 minutes]:
sync.runtime_SemacquireMutex(0x5557d0292f56?, 0xa0?, 0xa3?)
	/usr/local/go/src/runtime/sema.go:77 +0x26
sync.(*Mutex).lockSlow(0xc00a2a1800)
	/usr/local/go/src/sync/mutex.go:171 +0x165
sync.(*Mutex).Lock(...)
	/usr/local/go/src/sync/mutex.go:90
github.com/docker/docker/pkg/broadcaster.(*Unbuffered).Clean(0xc00a2a1800)
	/go/src/github.com/docker/docker/pkg/broadcaster/unbuffered.go:42 +0x3a
github.com/docker/docker/container/stream.(*Config).CloseStreams(0xc009e59a40)
	/go/src/github.com/docker/docker/container/stream/streams.go:102 +0x105
github.com/docker/docker/container.(*rio).Close(0xc00900a558)
	/go/src/github.com/docker/docker/container/container.go:806 +0x37
github.com/docker/docker/vendor/github.com/containerd/containerd.(*task).Delete(0xc00107b540, {0x5557d11f8148, 0xc0097176e0}, {0x0, 0x0, 0x5557d265df40?})
	/go/src/github.com/docker/docker/vendor/github.com/containerd/containerd/task.go:328 +0x2ab
```

are still going to block.

Reply:
You're right. However, my testing shows that it is even more complicated than that: during my testing `libcontainerd/remote.(*task).Delete()` calls `t.io.Wait()` [here](https://github.com/moby/moby/blob/4d67489/vendor/github.com/containerd/containerd/task.go#L339)
this is a call to [rio.Wait()](https://github.com/moby/moby/blob/4d674897f3748a28347f2233237dbfae7c18cdc1/container/container.go#L809)
this, in turn, calls `i.sc.Wait(context.Background())` - which hangs, probably because containerd-shim does not exit, probably because it waits for the receiver to read the data it sends (container's output that we do not read because attached reader is not reading). Maybe `rio.Wait()` should also accept a context and pass that to `i.sc.Wait()`

Reply:
After some more thinking, I see the following ways to change `handleContainerExit` to fix this blocking:
* move the deletion after waiting and closing IO, [like this](https://gist.github.com/asterite3/691d439187e11c6bfac4edf82e2c2f3e). This is the simplest fix and it seems to be in the spirit of containerd commit [fc535efe5845aa99f98b7fac58526b1b06c102ab](https://github.com/containerd/containerd/commit/fc535efe5845aa99f98b7fac58526b1b06c102ab) (see my other comment below for more details about that commit). This will give contanerd shim around 2 seconds to transfer the remaining container output to docker daemon.
* A more conservative approach: call `tsk.Delete(ctx)` *before* closing the IO, but, instead of waiting indefinitely for that call to return, only wait for some time, and close IO streams then ( `tsk.Delete(ctx)`  can be called in a separate goroutine for that). This second option makes the code of `handleContainerExit` larger and more complicated, but in this case the code will behave closer to how it was before the fix
 
According to my testing, both ways fix the problem
I can't say I fully understand the implications of the first option because I'm not super familiar with Docker architecture, whether it will work OK in all cases. I myself did not find other consequences beside the descreased timeout for IO completion (timeouts of contexts passed to `StreamConfig.Wait(ctx)` and `c.Reset(ctx, false)` can, of course, be increased to mitigate that). It would be great to get someone else's opinion here

Reply:
I have added a commit that implements the second, more conservative, approach. It uses a timeout of 3 seconds while waiting for `tsk.Delete(ctx)` before proceeding to closure of IO streams. In addition with a 2-second timeout passed to `StreamConfig.Wait`, dockerd will give around 5 seconds for container output to be transferred from containerd to readers. If we set this timeout to 4 or more seconds, `docker kill` will start complaining that container did not exit in 10 seconds and that it sends a "direct SIGKILL". Maybe that's OK but is the reason I chose a 3-second time and not more.

It can be noted that there is a context with 30-second timeout passed to `tsk.Delete(ctx)` and we could wait for that to expire, but my experiments showed that, if that context is cancelled, `tsk.Delete(ctx)` fails and containerd shim is not deleted, it remains running without a child process (seemingly forever).

Reply:
A different approach to fix this is to address the issue on containerd side. It is evident that, although `tsk.Delete` accepts a context, its cancellation is not fully respected and [`t.io.Wait()`](https://github.com/containerd/containerd/blob/2aeac9a/task.go#L337) may block forever. I'm not sure how the fix on containerd should be implemented though.

Interface `cio.IO` does not support passing contexts to its method calls, so we would have to either update that interface or add a method `WaitContext` to `rio` that's not in that interface and test for that in `Delete` using a type assertion.
Also, if I'm not mistaken, if `ctx` passed to `Delete` is cancelled, an RPC call `t.client.TaskService().Delete` would fail. So maybe `Delete` should pass another context, with lesser timeout, to that waiting method.
(And obviously this would require a PR to containerd and synchronizing the vendored version)